validation: test cgroups with different input values

dongsupark opened the pull request:
This PR tries to cover more test cases by running cgroup tests with different combinations of input values. Also it tries to print out detailed output results in the TAP format.

To do that, first of all, we change prototype of `RuntimeOutsideValidate()` as well as `AfterFunc()`, updating all corresponding call sites. And then we update cgroups tests. Changed cgroups tests are:

* Memory 
* CPU
* Network
* Blkio
* HugeTLB

Especially I needed to update CPU cgroups tests a lot, by following the same semantics of memory cgroups. For that, I had to add helpers in `validation/util/linux_resources_cpus.go`.

/cc @alban

liangchenye commented on generate/config.go:
```diff
@@ -94,7 +94,8 @@ func (g *Generator) initConfigLinuxResourcesBlockIO() {
 
-func (g *Generator) initConfigLinuxResourcesCPU() {
+// InitConfigLinuxResourcesCPU initializes CPU of Linux resources
+func (g *Generator) InitConfigLinuxResourcesCPU() {
```
this change LGTM

liangchenye commented on validation/linux_cgroups_blkio.go:
```diff
@@ -1,30 +1,100 @@
+	t.Header(0)
+	defer t.AutoPlan()
+
+	// It's assumed that a device /dev/sda (8:0) exists on the test system.
```
why do we need this assumption?

dongsupark replied:
@liangchenye 
I'm not sure why, as I'm not the original author of the blkio test. The test has always relied on static major/minor numbers, since the beginning:
https://github.com/opencontainers/runtime-tools/commit/f1e02ff6b3bca42a9fa33f918f55cf97da3ab193

What I'm doing with this PR is only to improve the existing test. Though during running the blkio test on my local system, I realized that the test fails because of a missing device `/dev/sda` (8:0). The local machine has only `/dev/nvme` (243:0), not `/dev/sda`. That's why I added the comment above.

Ideally we could improve this part so that it detects the root device at runtime, instead of hard-coding like that. Though I think that it should be done in a separate PR. If you want me to change so, I can make another PR.

liangchenye commented on validation/linux_cgroups_blkio.go:
```diff
@@ -1,30 +1,100 @@
+	return nil
+}
+
+func testEmptyBlkio() error {
```
I think it could be part of testBlkioCgroups. We can add more args to the ```cases struct```.  Take https://github.com/opencontainers/runtime-tools/blob/master/validation/state.go#L29 for example.

dongsupark replied:
@liangchenye That sounds like a good idea. Will try to do that.

liangchenye commented:
>> This PR tries to cover more test cases by running cgroup tests with different combinations of input values. Also it tries to print out detailed output results in the TAP format.

The `combinations` idea looks cool. I just wonder if the testing values were representative since we are not doing performance/pressure test.

dongsupark commented:
@liangchenye Sure, in some cases testing values could not be optimal. But I don't have a better idea.

dongsupark commented:
Rebased on top of master, and updated CPU & blkio tests.

* Added more nil checks in CPU tests
* Squashed empty blkio tests into `testBlkioCgroups`.

dongsupark commented:
Updated the PR, to replace `util.Fatal` with `t.Fail`, as done in https://github.com/opencontainers/runtime-tools/pull/645.

q384566678 commented on validation/util/linux_resources_devices.go:
```diff
@@ -48,6 +45,5 @@ func ValidateLinuxResourcesDevices(config *rspec.Spec, state *rspec.State) error
 		}
 	}
 
-	t.AutoPlan()
```
According to your idea, there is `t.AutoPlan()` in this file that has not been deleted yet.

dongsupark replied:
@q384566678 Good catch.
I've removed `t.AutoPlan()` from `linux_resources_devices.go` as well as `linux_resources_blkio.go`.

q384566678 commented:
Looks good, WDYT @liangchenye .

q384566678 commented:
LGTM
